test/testlib/OpenSSL/Test.pm

```diff
@@ -277,54 +276,106 @@ An example:
 
 =back
 
+=begin comment
+
+One might wonder over the complexity of C<apps>, C<fuzz>, C<test>, ...
+with all the lazy evaluations and all that.  The reason for this is that
+we want to make sure the directory in which those programs are found are
+correct at the time these commands are used.  Consider the following code
+snippet:
+
+  my $cmd = app(["openssl", ...]);
+
+  indir "foo", sub {
+      ok(run($cmd), "Testing foo")
+  };
+
+If there wasn't this lazy evaluation, the directory where C<openssl> is
+found would be incorrect at the time C<run> is called, because it was
+calculated before we moved into the directory "foo".
+
+=end comment
+
 =cut
 
-sub app {
+sub cmd {
     my $cmd = shift;
     my %opts = @_;
     return sub { my $num = shift;
-		 return __build_cmd($num, \&__apps_file, $cmd, %opts); }
+                 # Make a copy to not destroy the caller's array
+                 my @cmdargs = ( @$cmd );
+                 my @prog = __wrap_cmd(shift @cmdargs,
+                                       $opts{exe_shell} // ());
+
+                 return __decorate_cmd($num, [ @prog, quotify(@cmdargs) ],
+                                       %opts); }
```

Comment:
I'm not saying its syntactically wrong...just odd. I am used to seeing the close brace on a line on its own (and the opening brace with a newline after it)...maybe its because that's our C style and I'm just applying the same thinking to perl.